### PR TITLE
[RFC] net/netstats L1/L2 per neighbor transfer statistics

### PR DESCRIPTION
**Pull request by bergzand**
This WIP adds functionality for recording statistics split out per peer. The goal of these statistics are for RPL parent selection, for example based on the ETX of the connection to the potential parent.

A struct with these statistics is defined. At this moment an array of length 8 is added to the netdev struct. I went for this approach because ETX statistics can be radio dependent and if a device has multiple radios these statistics are clearly dependent on the radio. Statistics are updated on packets received from the neighbor or transmission to the neighbor.

Because the netdev callback used to process transmission result has no knowledge of the transmitted frame itself, I decided to implement a small ring buffer like construction to link transfers to the callback. This ring buffer assumes that the order of packets send to the radio and the callbacks are in the same order. The ring buffer consists of a array of netstats_peer_t pointers and 2 indexes for reading/writing positions. Multicast/broadcast transmissions are recorded as NULL ptr's preserving the order of this buffer.

RSSI and LQI values are kept as exponential weighted moving averages, this to have some average function over this data while keeping the stored data small. I prefer to store the RSSI values in dBm, but the current radio's return the raw values instead of dBm converted values.

I've added a netopt call to retrieve the statistics. I've used this to extend the `ifconfig` tool with these statistics.

This PR is a work in progress. It's probably not feasible to have this merged before the feature deadline.

I would like some feedback on the architecture build for this. If a different approach is preferred, I can rewrite this. Also, if a different algorithm for certain parts is preferred, I'd also like to hear

Todo:
- [x] ETX calculations
- [x] Implement byte counters
- [x] Some kind of freshness indication
- [x] RSSI in dBm, Needs a different PR to fix this in the radio drivers

### CODE REVIEW

**Comment by bergzand** [comment]
Rebased and resolved merge conflict

**Comment by bergzand** [comment]
Rework of implementation, including a rename.

Now based on #7447 and #7448, works better with #6895.
ETX and freshness is implemented.

The byte and packet counters are under an extra pseudomodule because they are not necessary for application requiring ETX such as RPL MRHOF.

**Comment by bergzand** [comment]
No longer WIP.

**Comment by bergzand** [comment]
Prevents a segfault when no empty/stale entries are present and added support for ethernet devices.

Note that ETX doesn't really make sense with tap because there is no L2 ack mechanism, but it simplifies testing a lot.

**Comment by smlng** [comment]
it looks like (pseudomodule) `netstats_neighbor_ext` depends on module `netstats_neighbor`, hence there should be an entry in `Makefile.dep` like:

```
ifneq (,$(filter netstats_neighbor_%, $(USEMODULE)))
  USEMODULE += netstats_neighbor
endif
```

@miri64 what do you think?

**Comment by bergzand** [comment]
Dependency added.

I just realized that this probably need some guards against the situation that L2 acks are disabled as most of the information is bogus then. The only correct information then is the tx counter and the rx information. Not recording any tx statistics might be the better option when L2 acks are disabled. I guess at least some documentation on this situation is preferred?

**Comment by bergzand** [comment]
rebased on master and removed commits that were no real dependencies.

ETX functionality testing requires either #7484 for the at86rf23(2|3) or #7448 for the mrf24j40.

**Comment by kYc0o** [comment]
Rebase please :)

**Comment by bergzand** [comment]
@kYc0o I want to refactor my code a bit, I found out RIOT has a CIB library, so my own implementation included in this PR is unnecessary.

I should have time for this small refactor and the rebase this week.

**Comment by bergzand** [comment]
Rebased and using `cib.h` instead of my own solution.

The changes in `netstats.h` are moved to `net/netstats/neighbor.h`. The main problem with this is that it creates some circular includes between `net/netdev.h` and `net/netstats/neighbor.h`.

I still have to test if the inclusion of `cib.h` actually works in practice, but I don't expect any real problems there.

**Comment by bergzand** [comment]
addressed comments from @cgundogan from #7450 ("waiting for other PR" tag was missing there).

File: sys/include/net/netstats.h [hunk 1]
```diff
@@ -53,6 +65,27 @@ typedef struct {
     uint32_t rx_bytes;          /**< received bytes */
 } netstats_t;
 
+/**
+ * @brief       Stats per peer struct
+ */
+typedef struct netstats_nb {
+    uint8_t l2_addr[8];     /**< Link layer address of the neighbor */
```
Comment (cgundogan) [inline sys/include/net/netstats.h]: hmm can I haz macro?
```c
#ifndef NETSTATS_ADDR_MAXLEN
#define NETSTATS_ADDR_MAXLEN (8)
#endif
```

Comment (cgundogan) [reply sys/include/net/netstats.h]: Oh, you already have somthing similar in:
sys/include/net/netstats/neighbor.h

Maybe you can join them?

File: sys/include/net/netstats/neighbor.h [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    NETSTATS_NB_BUSY,       /**< Failed due to medium busy */
+    NETSTATS_NB_NOACK,      /**< Failed due to no ack received */
+    NETSTATS_NB_SUCCESS,    /**< Succesful transmission */
+} netstats_nb_result_t;
+/** @} */
+
+/**
+ * @brief Max length of a L2 address
+ */
+#define NETSTATS_NB_L2_ADDR_MAX_SIZE      (8)
```
Comment (cgundogan) [inline sys/include/net/netstats/neighbor.h]:
```c
#ifndef NETSTATS_NB_L2_ADDR_MAX_SIZE
#define NETSTATS_NB_L2_ADDR_MAX_SIZE      (8)
#endif
```

File: sys/include/net/netstats/neighbor.h [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+
+/**
+ * @name @ref EWMA parameters
+ * @{
+ */
+#define NETSTATS_NB_EWMA_SCALE            100   /**< Multiplication factor of the EWMA */
```
Comment (cgundogan) [inline sys/include/net/netstats/neighbor.h]: Not sure if there's value to make all these defines configurable on compile time ..
Wouldn't hurt to wrap them around `#ifndef ..#endif`?

File: sys/include/net/netstats/neighbor.h [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+ *
+ * @param[in] dev          ptr to netdev device
+ * @param[in] l2_addr ptr  to the L2 address
+ * @param[in] l2_addr_len  length of the L2 address
+ * @param[in] rssi         RSSI of the received transmission in abs([dBm])
+ * @param[in] lqi          Link Quality Indication provided by the radio
+ *
+ * @return ptr to the updated record
+ */
+netstats_nb_t *netstats_nb_update_rx(netdev_t *dev, const uint8_t *l2_addr, uint8_t l2_addr_len, uint8_t rssi, uint8_t lqi);
```
Comment (cgundogan) [inline sys/include/net/netstats/neighbor.h]: pretty long line here

**Comment by bergzand** [comment]
@cgundogan Thanks for the initial review. As for testing, after the whole netif2 project is done, I can add some tests to this. For manual testing, #7449 (closed due to netif2 incompatibility) add output of these statistics to the ifconfig command.

File: sys/include/net/netstats/neighbor.h (hunk 3 above)
Comment (bergzand) [reply sys/include/net/netstats/neighbor.h]: Most of these make sense to have configurable. All of them control the timings in some way. For a mostly static network, timings can probably be a bit longer and for more dynamic you probably want some shorter timings and a more responsive averaging, maybe even a higher noack penalty.

TL;DR: Yes.

**Comment by bergzand** [comment]
After some thinking about the current structure and how to port it to netif2 I came to a number of things I don't like about it:

### netstats_nb shouldn't have a dependency on netdev.

Currently a netdev struct is used as an argument in a number of functions. This creates an unecessary dependency on netdev. As the code shouldn't require anything from the actual network interface struct, this dependency is IMHO wrong. It is only used as a way to access the array of neighbor statistics, so this whole issue could be resolved by either having those functions rewritting to get the first statistics struct from the array as a pointer and work from there or encapsulate the whole array in a separate struct and pass those around.

### netstats_nb shouldn't have to solve the issue of matching transmission statistics to the original packets.

This is required because the current netdev/gnrc implementation "forgets" a packet as soon as the radio is instructed to transmit it. So at the moment the transmit statistics are received from the callback, it is not known anymore what the destination MAC is. Now IMHO this issue of matching transmission results and destination MAC is not something that the statistics module should have to worry about. I think it is better to move this to a separate netdev part (I'm looking at you #7736) and keep the netstats_nb module about statistics and statistics averaging only. Maybe a small `netdev_layered_t` module that remembers the L2 pktsnip (or parts of this information) until after a transmission might help here.

### Either the statistics should be in gnrc_netif and handled in gnrc_netif or contained in netdev and handled in netdev:

At this moment, the array and CIB information is contained in the `netdev_t` struct of the radio, but all the actual handling ~is~ was done in the gnrc_netdev code. This creates a weird mix between layers which I consider ugly. For now I'm considering adding the matching to netdev and the statistics themselves to gnrc_netif.

**Comment by smlng** [comment]
as long as this is tagged RFC, the milestone seems premature.

**Comment by miri64** [comment]
@bergzand I assume you are planning to port this to #8198 in the future?

**Comment by bergzand** [comment]
Yes, At some point I hope to build a v2 of this based on #8198

**Comment by tcschmidt** [comment]
@bergzand #8198 is merged, how to adapt right now?

**Comment by bergzand** [comment]
@tcschmidt As requested [here](https://github.com/RIOT-OS/RIOT/pull/8198#issuecomment-374007819) #8198 contained only the initial headers and code required for the layered approach. As long as the netdev stack is not adapted to actually use the code, it is not possible to reworking this PR in a useful way.

**Comment by benpicco** [comment]
@bergzand is this still true?
What needs to be done before it's possible to add neighbor statistics into the stack and where should they go now?

**Comment by miri64** [comment]
Ping @bergzand?

**Comment by bergzand** [comment]
@miri64 @benpicco Unfortunately I don't have time to work on this. I'm also not sure where this functionality should be hooked into the stack. Mostly as there are now multiple ideas floating around on how the link/phy layer should evolve.

**Comment by chrysn** [comment]
If this were resolved, would there be much work left to do to implement MRHOF? (The contiki users around me are kind of reluctant to switch that off).

**Comment by bergzand** [comment]
> If this were resolved, would there be much work left to do to implement MRHOF? (The contiki users around me are kind of reluctant to switch that off).

[Around 350 lines of code](https://github.com/RIOT-OS/RIOT/pull/7450/files/db04c332d0a2345aca030d3bdef98da5767e3778..d18ca58905a6a493cb790aa025116f167929d297) depending a bit on how much things changed in two years

**Comment by benpicco** [comment]
> netstats_nb shouldn't have a dependency on netdev.

This might actually be useful.
I think there is a case for allowing device-specific extra information per-neighbor.
E.g. with `at86rf215` you might want to use different data rates for different nodes, depending on how good the link quality is.
You could also talk to MR-O-QPSK nodes and legacy O-QPSK nodes simultaneously if there was a way to store which node used which modulation.

**Comment by miri64** [comment]
Would have thought the bot removes the stale label if you re-open the issue...

**Comment by benpicco** [comment]
> @miri64 @benpicco Unfortunately I don't have time to work on this. I'm also not sure where this functionality should be hooked into the stack. Mostly as there are now multiple ideas floating around on how the link/phy layer should evolve.

I can take over. From what I gathered by talking to @miri64 was that it should not be part of GNRC (so not extending the NIB) but rather a separate L2 database - so pretty much what you did here.

File: sys/net/netstats/netstats_neighbor.c [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+void netstats_nb_create(netstats_nb_t *entry, const uint8_t *l2_addr, uint8_t l2_len)
+{
+    memset(entry, 0, sizeof(netstats_nb_t));
+    memcpy(entry->l2_addr, l2_addr, l2_len);
+    entry->l2_addr_len = l2_len;
+    entry->etx = NETSTATS_NB_ETX_INIT * NETSTATS_NB_ETX_DIVISOR;
+}
+
+/* find the oldest inactive entry to replace. Empty entries are infinity old */
+netstats_nb_t *netstats_nb_get_or_create(netdev_t *dev, const uint8_t *l2_addr, uint8_t len)
```
Comment (benpicco) [inline sys/net/netstats/netstats_neighbor.c:44]: Wouldn't it be better to evict the entry with the worst signal?

File: sys/include/net/netstats.h [hunk 6]
```diff
@@ -53,6 +65,27 @@ typedef struct {
...
+    uint8_t l2_addr_len;    /**< Length of netstats_nb::l2_addr */
+    uint16_t etx;           /**< ETX of this peer */
+#ifdef MODULE_NETSTATS_NEIGHBOR_EXT
+    uint8_t rssi;           /**< Average RSSI of received frames in abs([dBm]) */
+    uint8_t lqi;            /**< Average LQI of received frames */
+    uint32_t tx_count;      /**< Number of sent frames to this peer */
+    uint32_t tx_failed;     /**< Number of failed transmission tries to this peer */
+    uint32_t rx_count;      /**< Number of received frames */
+    uint32_t tx_bytes;      /**< Bytes sent */
+    uint32_t rx_bytes;      /**< Bytes received */
```
Comment (benpicco) [inline sys/include/net/netstats.h:82]: If those are just for printing out information, I'd rather drop them - 20 bytes / neighbor is quite a lot to keep around for curiosity's sake.

**Comment by miri64** [comment]
Very timely, stale bot, very timely ;-).
